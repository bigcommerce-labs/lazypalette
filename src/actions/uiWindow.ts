import { ColorResult } from 'react-color';
import uuid from 'uuid';

export enum UIWindowActionTypes {
    OPEN_UI_WINDOW = 'OPEN_UI_WINDOW',
    CLOSE_UI_WINDOW = 'CLOSE_UI_WINDOW',
}

export enum UIWindowTypes {
    COLOR_PICKER = 'COLOR_PICKER',
}

export interface UIWindowContent {
    position: { x: number, y: number };
}

export interface ColorPickerContent extends UIWindowContent {
    color: string;
    onChange?(value: ColorResult): void;
}

export interface OpenUIWindowAction<Content> {
    payload: {
        content: Content;
        id: string;
        type: UIWindowTypes;
    };
    type: UIWindowActionTypes.OPEN_UI_WINDOW;
}

export interface CloseUIWindowAction {
    payload: {
        id: string,
    };
    type: UIWindowActionTypes.CLOSE_UI_WINDOW;
}

export interface OpenColorPickerArgs {
    color: string;
    id?: string;
    position: { x: number, y: number };
    onChange?(value: ColorResult): void;
}

export function openColorPicker(content: OpenColorPickerArgs): OpenUIWindowAction<ColorPickerContent> {
    return {
        payload: {
            content: {
                color: content.color,
                onChange: content.onChange,
                position: content.position,
            },
            id: content.id || uuid(),
            type: UIWindowTypes.COLOR_PICKER,
        },
        type: UIWindowActionTypes.OPEN_UI_WINDOW,
    };
}

export function closeUIWindow(id: string): CloseUIWindowAction {
    return {
        payload: { id },
        type: UIWindowActionTypes.CLOSE_UI_WINDOW,
    };
}
