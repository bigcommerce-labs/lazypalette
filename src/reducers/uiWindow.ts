import { Action } from '../actions/action';
import { UIWindowActionTypes, UIWindowContent, UIWindowTypes } from '../actions/uiWindow';

export interface UIWindowData {
    content: UIWindowContent;
    id: string;
    type: UIWindowTypes;
}

export interface UIWindowState {
    uiWindows: UIWindowData[];
}

const initialState: UIWindowState = {
    uiWindows: [],
};

function uiWindow(state: UIWindowState = initialState, action: Action): UIWindowState {
    if (action.error) {
        return state;
    }

    switch (action.type) {
        case UIWindowActionTypes.OPEN_UI_WINDOW:
            return { ...state,
                uiWindows: [...state.uiWindows,
                    {
                        content: action.payload.content,
                        id: action.payload.id,
                        type: action.payload.type,
                    }],
            };
        case UIWindowActionTypes.CLOSE_UI_WINDOW:
            return { ...state, uiWindows: state.uiWindows.filter(m => m.id !== action.payload.id) };
        default:
            return state;
    }
}

export default uiWindow;
