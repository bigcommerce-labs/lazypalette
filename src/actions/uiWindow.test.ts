import * as uiWindowActions from './uiWindow';

describe('uiWindow actions', () => {
    describe('when we create an openColorPicker action', () => {
        it('creates the correct action', () => {
            const action = uiWindowActions.openColorPicker({
                color: 'FFFFFF',
                id: '1234',
                onChange: () => (1),
                position: { x: 10, y: 15 },
            });

            expect(action).toMatchSnapshot();
        });
    });
    describe('when we create a closeUIWindow action', () => {
        it('creates the correct action', () => {
            const action = uiWindowActions.closeUIWindow('5678');

            expect(action).toMatchSnapshot();
        });
    });
});
