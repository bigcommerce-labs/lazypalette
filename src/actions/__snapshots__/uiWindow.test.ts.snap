// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uiWindow actions when we create a closeUIWindow action creates the correct action 1`] = `
Object {
  "payload": Object {
    "id": "5678",
  },
  "type": "CLOSE_UI_WINDOW",
}
`;

exports[`uiWindow actions when we create an openColorPicker action creates the correct action 1`] = `
Object {
  "payload": Object {
    "content": Object {
      "color": "FFFFFF",
      "onChange": [Function],
      "position": Object {
        "x": 10,
        "y": 15,
      },
    },
    "id": "1234",
    "type": "COLOR_PICKER",
  },
  "type": "OPEN_UI_WINDOW",
}
`;
